File: server_kotlin/src/main/kotlin/classes/Message.kt
package com.perceivers25.betalk.classes

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.*

@Serializable
class Message(
    val id: String,
    val author: String?,
    val textContent: String?,
    val fileContent: ByteArray?,
    val time: String,
    val type: MessageType,
) {
    init {
        assert(textContent != null || fileContent != null)
    }

    @Serializable
    enum class MessageType {
        @SerialName("USER_MESSAGE")
        UserMessage,

        @SerialName("SYSTEM_MESSAGE")
        SystemMessage,
    }

    companion object {
        fun newSystemMessage(messageTextContent: String): Message {
            return Message(
                UUID.randomUUID().toString(), null, messageTextContent, null, DateTimeFormatter
                    .ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS")
                    .withZone(ZoneOffset.UTC)
                    .format(Instant.now()), MessageType.SystemMessage
            )
        }
    }
}

File: server_rust/src/models/message.rs
use anyhow::{Result, anyhow};
use chrono::DateTime;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MessageType {
  UserMessage,
  SystemMessage,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Message {
  pub id: String,
  pub author: Option<String>,
  pub text_content: Option<String>,
  pub file_content: Option<Vec<u8>>,
  pub time: DateTime<chrono::Utc>,
  #[serde(rename = "type")]
  pub message_type: MessageType,
}

impl Message {
  pub fn new_system_message(text_content: String) -> Self {
    Self {
      id: uuid::Uuid::new_v4().to_string(),
      author: None,
      text_content: Some(text_content),
      file_content: None,
      time: chrono::Utc::now(),
      message_type: MessageType::SystemMessage,
    }
  }

  pub fn assert(&self) -> Result<()> {
    if self.text_content.is_none() && self.file_content.is_none() {
      return Err(anyhow!("Message must contain either text or file content"));
    }

    Ok(())
  }
}
